### PR TITLE
Blog about pod scheduling readiness

### PR DESCRIPTION
**Pull request by Huang-Wei**
~Placeholder~ PR for the blog "Pod Scheduling Readiness".

/sig scheduling

KEP: https://github.com/kubernetes/enhancements/issues/3521

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP] Add blog article about pod scheduling readiness

**Comment by bradmccoydev** [comment]
Hello @Huang-Wei!  v1.26 Comms shadow here. This feature blog is on a feature tracked for release, the deadline for submitting a draft is the 29th of November; this should be considered the hard limit since we will need to review/edit/discuss the draft, so if at all possible it's better to submit it earlier to avoid any problems.

Any doubts, we're here to help! cc @fsmunoz

**Comment by Huang-Wei** [comment]
@bradmccoydev thanks for the reminder. I will have some cycles early next week, and it's quite optimistic to finish the draft before the deadline.

**Comment by fsmunoz** [comment]
Hello @Huang-Wei , today is Review Readiness deadline, it's important not to miss it since it can imply dropping the feature blog :( If we can be of any help, we'e here!

**Comment by Huang-Wei** [comment]
@fsmunoz it's ready for review now.

(cc @ahg-g )

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+---
+layout: blog
+title: "Kubernetes 1.26: Pod Scheduling Readiness to Alpha"
+date: 2022-11-29
+slug: pod-scheduling-readiness-alpha-1-26
+---
+
+**Author:** Wei Huang (Apple), Abdullah Gharaibeh (Google)
+
+Kubernetes 1.26 introduced a new Pod feature named SchedulingGates. SchedulingGates are keys that
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:10]: (nit)
```suggestion
Kubernetes 1.26 introduced a new Pod feature: _scheduling gates_. In Kubernetes, scheduling gates are keys that
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+```
+NAME       READY   STATUS            RESTARTS   AGE
+test-pod   0/1     SchedulingGated   0          10s
+```
+
+To clear the gates, you update the Pod by removing the `schedulingGates` string literals.
+The gates do not need to be removed all at once, but only when all the gates are removed the
+scheduler will start to consider the Pod for scheduling.
+
+Under the hood, SchedulingGates is implemented as a PreEnqueue scheduler plugin, a new scheduler
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:48]:
```suggestion
Under the hood, scheduling gates are implemented as a PreEnqueue scheduler plugin, a new scheduler
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+and therefore saves unnecessary scheduling attempts. Clearing the gates is the responsibility of
+external controllers with knowledge of when the Pod should be considered for scheduling (e.g., 
+a resource provisioner).
+
+![](scheduler-workflow-2.png)
+
+## How does it work?
+
+SchedulingGates in general works very similar to Finalzers. Pods with non-empty 
+`spec.schedulingGates` field will carry the status `SchedulingGated` and will be blocked from
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:35]: Is this right:
```suggestion
`spec.schedulingGates` field will gain the `SchedulingGated`
[condition](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) and will be blocked from
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+scheduling. Note that more than one gate can be added, but they all should be added upon Pod
+creation (e.g., you can add them as part of the spec of via a mutating webhook).
+
+```
+NAME       READY   STATUS            RESTARTS   AGE
+test-pod   0/1     SchedulingGated   0          10s
+```
+
+To clear the gates, you update the Pod by removing the `schedulingGates` string literals.
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:44]:
```suggestion
To clear the gates, you update the Pod by removing all of the items from the Pod's `schedulingGates` field.
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+An important use case this feature enables is dynamic quota management. Kubernetes supports
+ResourceQuota, however quotas are enforced by the api-server on Pod creation. For example, if a
+new Pod exceeds the cpu quota, it gets rejected. The api-server doesn't queue the Pod; therefore,
+whoever created the Pod needs to continuously attempt to recreate it again!
+
+SchedulingGates allows an external quota manager to address the above limitation of ResourceQuota.
+Specifically, the manager could add a "quotaCheck" scheduling gate to all Pods created in the
+cluster (using a mutating webhook). The manager would then remove the gate when there is quota to
+start the Pod.
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:61]: I'd expect to see:
```suggestion
Scheduling gates allow an external quota manager to address the above limitation of ResourceQuota.
Specifically, the manager could add a `example-com/quota-check` scheduling gate to all Pods created in the
cluster (using a mutating webhook). The manager would then remove the gate when there is quota to
start the Pod.
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+The gates do not need to be removed all at once, but only when all the gates are removed the
+scheduler will start to consider the Pod for scheduling.
+
+Under the hood, SchedulingGates is implemented as a PreEnqueue scheduler plugin, a new scheduler
+framework extension point that is invoked at the beginning of each scheduling cycle.
+
+## Use Cases
+
+An important use case this feature enables is dynamic quota management. Kubernetes supports
+ResourceQuota, however quotas are enforced by the api-server on Pod creation. For example, if a
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:54]:
```suggestion
[ResourceQuota](/docs/concepts/policy/resource-quotas/), however the API server enforces quota
at the time you attempt Pod creation. For example, if a
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+new Pod exceeds the cpu quota, it gets rejected. The api-server doesn't queue the Pod; therefore,
+whoever created the Pod needs to continuously attempt to recreate it again!
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:56]:
```suggestion
new Pod exceeds the CPU quota, it gets rejected. The API server doesn't queue the
Pod; therefore, whoever created the Pod needs to continuously attempt to recreate it again.
This either means a delay between resources becoming available and the Pod actually running,
or it means load on the API server as the client constantly makes new attempts.
```

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+SchedulingGates allows an external quota manager to address the above limitation of ResourceQuota.
+Specifically, the manager could add a "quotaCheck" scheduling gate to all Pods created in the
+cluster (using a mutating webhook). The manager would then remove the gate when there is quota to
+start the Pod.
+
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:62]:
```suggestion
A timeout or garbage collection mechanism would round off this example, where
Pods that have been queued for more than (say) 24 hours are deleted, and any
owning controller can request creating a replacement.

```
?

(not sure if this is reasonable / supported)

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+## Whats next?
+
+To use this feature, the `PodSchedulingReadiness` feature gate must be enabled in the API Server
+and Scheduler. You're more than welcome to test it out and tell us (sig-scheduling) what you think!
+
+## Additional resources
+
+- [Official PodSchedulingReadiness docs](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:70]:
```suggestion
- [Pod Scheduling Readiness](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/) in the Kubernetes documentation
```

**Review by sftim — COMMENTED** [review]
Thanks.

If feasible, use SVG images - see the [diagram guide](https://kubernetes.io/docs/contribute/style/diagram-guide/) for our docs guidelines, which mostly also apply to blog articles.

https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/ has some more tips.

**Comment by sftim** [comment]
/hold

Ready for release comms to assign a publication date

**Comment by fsmunoz** [comment]
Hello, Comms Lead for v1.26 here: this Feature Blog PR is **confirmed for the post-release Feature Blog series**. We will add the publishing date in the next days. Thank you for your work!

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 3 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:35]: Not quite, as it's not a new condition. The condition itself is still of type `PodScheduled`, just with a new reason `SchedulingGated`. How about rewording as "... will show as status ..."?

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 7 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:56]: I'd like to reword like this:

```
... or it means load on the API server and Scheduler as the client constantly makes new attempts. ...
```

(for API Server, the load comes up continuous refreshing its failure condition; for Scheduler, it comes up unnecessary scheduling cycles)

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 8 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md:62]: It was discussed in the KEP discussion. The consensus was to hold any explicit timeout until getting enough user feedback. So maybe not mention it for now.

**Comment by Huang-Wei** [comment]
@sftim thanks for the detailed review. All wording comments have been addressed. Working on using Mermaid to redraw the pictures.

**Comment by Huang-Wei** [comment]
@sftim Mermaid bits updated. Although I tried pretty hard to make it close to my original pictures, it's somehow restricted on how different parts are positioned, and how the arrows get rendered. So take a look and let me know if it's good enough.

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
+---
+layout: blog
+title: "Kubernetes 1.26: Pod Scheduling Readiness to Alpha"
+date: 2022-11-29
+slug: pod-scheduling-readiness-alpha-1-26
+---
+
+**Author:** Wei Huang (Apple), Abdullah Gharaibeh (Google)
+
+Kubernetes 1.26 introduced a new Pod feature: _scheduling gates. In Kubernetes, scheduling gates
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md:10]:
```suggestion
Kubernetes 1.26 introduced a new Pod feature: _scheduling gates_. In Kubernetes, scheduling gates
```

**Comment by Huang-Wei** [comment]
> For the diagrams, if you want to, you can save to SVG and load that up into an editor such as Inkscape, then make any tweak you see as reasonable.

I prefer to use Mermaid texts only and thus it's editable in the future. For embedding a tweaked SVG, it's still not convenient to modify in the future.

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md (hunk 10 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md]: done.

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+ class pod plain;
+ class note note;
+ class Scheduler Scheduler;
+{{< /mermaid >}}
+
+Scheduling gates helps address this problem. It allows declaring that newly created Pods are not
+ready for scheduling. When scheduling gates are present on a Pod, the scheduler ignores the Pod
+and therefore saves unnecessary scheduling attempts. Clearing the gates is the responsibility of
+external controllers with knowledge of when the Pod should be considered for scheduling (e.g., 
+a resource provisioner).
```
Comment (ahg-g) [inline content/en/blog/_posts/2022-11-29-pod-scheduling-readiness.md:52]:
```suggestion
Scheduling gates helps address this problem. It allows declaring that newly created Pods are not
ready for scheduling. When scheduling gates are present on a Pod, the scheduler ignores the Pod
and therefore saves unnecessary scheduling attempts. Those Pods will also be ignored by Cluster Autoscaler if you have it installed in the cluster. 

Clearing the gates is the responsibility of external controllers with knowledge of when the Pod should be considered for scheduling (e.g., a quota manager).
```

**Comment by fsmunoz** [comment]
Hello, Comms Lead for v1.26 here. **The publication order and date for the Feature Blog series has been finalised** and the [tracking board](https://github.com/orgs/kubernetes/projects/98/views/11?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=22229692&groupedBy%5BcolumnId%5D=15217749&filterQuery=enhancements-status%3ATracked+blog-status%3A%22Ready+for+Review%22) updated (cf. [discussion](https://github.com/kubernetes/sig-release/discussions/2105)).

The publication date for this article is **26-12-2022 (December 26)**. 

Thanks!

**Comment by Huang-Wei** [comment]
@fsmunoz thanks, the document was renamed to carry `2022-12-26` in its path.

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
+---
+layout: blog
+title: "Kubernetes 1.26: Pod Scheduling Readiness to Alpha"
+date: 2022-11-29
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:4]:
```suggestion
date: 2022-11-26
```

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
+---
+layout: blog
+title: "Kubernetes 1.26: Pod Scheduling Readiness to Alpha"
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:3]:
```suggestion
title: "Kubernetes 1.26: Pod Scheduling Readiness"
```
Features don't really graduate to alpha, as alpha is the first stage we tell end users about.

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+ class queue,sched_cycle,schedulable k8s;
+ class pod plain;
+ class note note;
+ class popout popout;
+ class Scheduler Scheduler;
+{{< /mermaid >}}
+
+## How does it work?
+
+Scheduling gates in general works very similar to Finalizers. Pods with non-empty
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:88]: (nit)
```suggestion
Scheduling gates in general works very similar to Finalizers. Pods with a non-empty
```

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+An important use case this feature enables is dynamic quota management. Kubernetes supports
+[ResourceQuota](/docs/concepts/policy/resource-quotas/), however the API Server enforces quota at
+the time you attempt Pod creation. For example, if a new Pod exceeds the CPU quota, it gets rejected.
+The API Server doesn't queue the Pod; therefore, whoever created the Pod needs to continuously attempt
+to recreate it again. This either means a delay between resources becoming available and the Pod
+actually running, or it means load on the API server and Scheduler due to constant attempts.
+
+Scheduling gates allows an external quota manager to address the above limitation of ResourceQuota.
+Specifically, the manager could add a `example-com/quota-check` scheduling gate to all Pods created in the
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:115]:
```suggestion
Specifically, the manager could add a `example.com/quota-check` scheduling gate to all Pods created in the
```
If `example.com/quota-check` isn't a valid scheduling gate name, is that actually something we'd want to fix for beta?

I'd expect an analogy with https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set and for similar motivations: different products that people might install onto a cluster can avoid clashes between keys.

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+cluster (using a mutating webhook). The manager would then remove the gate when there is quota to
+start the Pod.
+
+## Whats next?
+
+To use this feature, the `PodSchedulingReadiness` feature gate must be enabled in the API Server
+and Scheduler. You're more than welcome to test it out and tell us (sig-scheduling) what you think!
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:122]: (nit)
```suggestion
and scheduler. You're more than welcome to test it out and tell us (SIG Scheduling) what you think!
```

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+## Additional resources
+
+- [Pod Scheduling Readiness](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)
+in the Kubernetes documentation
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:127]: (nit)
```suggestion
  in the Kubernetes documentation
```

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+Kubernetes 1.26 introduced a new Pod feature: _scheduling gates_. In Kubernetes, scheduling gates
+are keys that tell the scheduler when a Pod is ready to be considered for scheduling.
+
+## What problem does it solve?
+
+When a Pod is created, the scheduler will continuously attempt to find a node that fits it. This
+infinite loop continues until the scheduler either finds a node for the Pod, or the Pod gets deleted.
+
+Pods that remain unschedulable for long periods of time (e.g., ones that are blocked on some external event) 
+waste scheduling cycles. A scheduling cycle may take ~20ms or more depending on the complexity of
```
Comment (sftim) [inline content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:19]: (nit)
```suggestion
waste scheduling cycles. A scheduling cycle may take ≅20ms or more depending on the complexity of
```

**Review by sftim — COMMENTED** [review]
Some more feedback

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md (hunk 15 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md:115]: For alpha impl., we don't enforce the name to be full qualified style. It's tracked for beta to evaluate: [#113608](https://github.com/kubernetes/kubernetes/issues/113608)

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md (hunk 12 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md]: updated to 2022-12-26

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 3 above)
Comment (sftim) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md]: > type `PodScheduled`, just with a new reason `SchedulingGated`

That sounds like good wording! You can also explain how `kubectl` represents this.

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 7 above)
Comment (sftim) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md]: Try this:
```markdown
or it means extra load for the API server and for the scheduler, due to the frequent,
repeated attempts to create a healthy Pod.
```

What do you think?

File: content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md (hunk 8 above)
Comment (sftim) [reply content/en/blog/_posts/2022-11-29-pod-scheduling-readiness/index.md]: How about saying
“for the alpha implementation, the manager is responsible for any garbage collection around
Pods that have been scheduling gated for too long. Kubernetes 1.26 itself doesn't implement
any mechanism for this”.

Is that fair?

File: content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md (hunk 15 above)
Comment (Huang-Wei) [reply content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md]: `-` was a typo... it's updated with `example.com/quota-check`.

Comment (sftim) [reply content/en/blog/_posts/2022-12-26-pod-scheduling-readiness.md]: Ah! All is clear. Thanks for explaining.

**Comment by fsmunoz** [comment]
Marking this Ready for Publishing in Comms tracking, thank you for all the work - minor edits/reviews are still doable until the release date.

/lgtm
